File: Exercises2024/Chapter06_vle_free_energy/Fortran/1-Widom/Source/mc_nvt.f
      Program Mc_Nvt
      Implicit None
C____________________________________________________C
C                                                    C
C                                                    C
C   Equation Of State Of The Lennard-Jones Fluid     C
C                                                    C
C____________________________________________________C

      Include 'potential.inc'
      Include 'parameter.inc'
      Include 'system.inc'
      Include 'conf.inc'

      Integer NumberOfEquilibrationCycles, NumberOfProductionCycles,
     &     SamplingFrequency, Mode, Icycl, NumberOfDisplacementsPerCycle,
     &     Attempts, Accepted, NumberOfCycles, Imove, J
      Double Precision RunningEnergy, TotalEnergy, RunningVirial,
     &     TotalVirial, MaxDisplacement, PressureSum, PressureCount, Pressure,
     &     ChemicalPotentialSum, ChemicalPotentialCount,Xi,Yi,Zi,
     &     EnergySum, EnergySquaredSum, EnergyCount, dummy1, dummy2, 
     &     RandomNumber
 
      Write (6, *) '**************** Mc_Nvt ***************'

C     ---Initialize Sysem

      Call Readdat(NumberOfEquilibrationCycles, NumberOfProductionCycles, 
     &     SamplingFrequency, NumberOfDisplacementsPerCycle, MaxDisplacement)

      PressureSum = 0.0d0
      PressureCount = 0.0d0
      EnergySum = 0.0d0
      EnergySquaredSum = 0.0d0
      ChemicalPotentialSum = 0.0d0
      ChemicalPotentialCount = 0.0d0

      Open (22,FILE='movie.pdb',FORM='FORMATTED',STATUS='REPLACE')

C     ---Total Energy Of The System

      Call EnergySystem(TotalEnergy, TotalVirial)
      Write (6, 99001) TotalEnergy, TotalVirial
      RunningEnergy = TotalEnergy
      RunningVirial = TotalVirial

C     ---Start Mc-Cycle

      Do Mode = 1, 2

C     --- Mode=1 Equilibration
C     --- Mode=2 Production
        
        If (Mode.Eq.1) Then
          NumberOfCycles = NumberOfEquilibrationCycles
          If (NumberOfCycles.Ne.0) Write (6, *) ' Start Equilibration '
        Else
          If (NumberOfCycles.Ne.0) Write (6, *) ' Start Production '
          NumberOfCycles = NumberOfProductionCycles
        End If
        
        Attempts = 0
        Accepted = 0
        
C     ---Intialize The Subroutine That Adjust The Maximum Displacement
        
        Call Adjust(Attempts, Accepted, MaxDisplacement)
        
        Do Icycl = 1, NumberOfCycles
          
          Do Imove = 1, NumberOfDisplacementsPerCycle
              
C     ---Attempt To Displace A Particle
              
            Call Mcmove(RunningEnergy, RunningVirial, Attempts, Accepted, 
     &           MaxDisplacement)
          End Do
            
          If (Mode.Eq.2) Then
              
C     ---Sample Averages
              
            If (Mod(Icycl,SamplingFrequency).Eq.0) Then
              Call Sample(Icycl, RunningEnergy, RunningVirial, Pressure)
              PressureSum = PressureSum + Pressure
              PressureCount = PressureCount + 1.0d0
              
              EnergySquaredSum=EnergySquaredSum+RunningEnergy*RunningEnergy
              EnergySum=EnergySum+RunningEnergy
              EnergyCount=EnergyCount+1.0
                
                
Ccccccccccccccccccccccccccccccccccccccccc
C     Calculate The Chemical Potential: C
C     Do 10 Trial Chains                C
C     Calculate The Average Of          C
C     [Exp(-Beta*Energy)]               C
C     You Can Use The Subroutine        C
C     EnergyParticle For This.          C
Ccccccccccccccccccccccccccccccccccccccccc
                
              Do J=1,10

C     Start Modification

                  
C     End   Modification
                  
              Enddo
            Endif
          Endif
            
          If(Mod(Icycl,20).Eq.0) Call Writepdb(22)

          If (Mod(Icycl,NumberOfCycles/5).Eq.0) Then
            Write (6, *) '======>> Done ', Icycl, ' Out Of ', NumberOfCycles
              
C     ---Write Intermediate Configuration To File
              
            Call Store(MaxDisplacement)
              
C     ---Adjust Maximum Displacements
              
            Call Adjust(Attempts, Accepted, MaxDisplacement)
          End If
        End Do
        If (NumberOfCycles.Ne.0) Then
          If (Attempts.Ne.0) Write (6, 99003) Attempts, Accepted, 
     &         100.0d0*Dble(Accepted)/Dble(Attempts)

C     ---Test Total Energy

          Call EnergySystem(TotalEnergy, TotalVirial)
          If (Abs(TotalEnergy-RunningEnergy).Gt.1.D-6) Then
            Write (6, *) 
     &             ' ######### Problems Energy ################ '
          End If
          If (Abs(TotalVirial-RunningVirial).Gt.1.D-6) Then
            Write (6, *) 
     &             ' ######### Problems Virial ################ '
          End If
          Write (6, 99002) TotalEnergy, RunningEnergy, 
     &         TotalEnergy - RunningEnergy, TotalVirial, 
     &         RunningVirial, TotalVirial - RunningVirial
          Write (6,*)
            
Ccccccccccccccccccccccccccccccccccccccccccccccc
Ccccccccccccccccccccccccccccccccccccccccccccccc
C     Print Chemical Potential And Pressure   C
Ccccccccccccccccccccccccccccccccccccccccccccccc
Ccccccccccccccccccccccccccccccccccccccccccccccc
            
          If(Mode.Eq.2) Then
              
            Write(6,*) 'Heat Capacity                      :',0.0
            Write (6,*) 'Average Pressure                  : ',
     &           PressureSum/PressureCount
            Write (6,*) 'Chemical Potential                : ',
     &           -Log((ChemicalPotentialSum/ChemicalPotentialCount)*
     &           (Box*Box*Box/Dble(NumberOfParticles)))/Beta
          End If
        End If
      End Do
      Call Store(MaxDisplacement)
      Close (22)
      Stop
        
99001 Format (' Total Energy Initial Configuration: ', F12.5, /, 
     &        ' Total Virial Initial Configuration: ', F12.5)
99002 Format (' Total Energy End Of Simulation    : ', F12.5, /, 
     &        '       Running Energy              : ', F12.5, /, 
     &        '       Difference                  :  ', E12.5, /, 
     &        ' Total Virial End Of Simulation    : ', F12.5, /, 
     &        '       Running Virial              : ', F12.5, /, 
     &        '       Difference                  :  ', E12.5)
99003 Format (' Number Of Att. To Displ. A Part.  : ', I10, /, 
     &        ' Success: ', I10, '(= ', F5.2, '%)')
      End

File: Exercises2024/Chapter06_vle_free_energy/C/1-Widom/Source/mc_nvt.c
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include "system.h"
#include "ran_uniform.h"


// equation of state of the Lennard-Jones fluid
 
int main(void)
{
  int i,j,k;
  int NumberOfCycles;
  double PressureSum,PressureCount,Pressure;
  double ChemicalPotentialSum,ChemicalPotentialCount,Dummy1,Dummy2;
  double EnergySquaredSum,EnergySum,EnergyCount;
  VECTOR pos;
  FILE *FilePtrMovie,*FilePtr;

  printf("**************** Mc_Nvt ***************\n");

  // initialize system
  ReadInputData();

  FilePtr=fopen("results.dat","w");
  FilePtrMovie=fopen("movie.pdb","w");

  NumberOfCycles=0;
  PressureSum=0.0;
  PressureCount=0.0;
  ChemicalPotentialSum=0.0;
  ChemicalPotentialCount=0.0;
  EnergySquaredSum=0.0;
  EnergySum=0.0;
  EnergyCount=0;

  // total energy of the system
  EnergySystem();
  printf(" Total Energy Initial Configuration: %lf\n",TotalEnergy);
  printf(" Total Virial Initial Configuration: %lf\n",TotalVirial);

  RunningEnergy=TotalEnergy;
  RunningVirial=TotalVirial;

  // start MC-cycles

  for(i=1;i<=2;i++)
  {
    // i=1 equilibration
    // i=2 production

    if (i==EQUILIBRATION)
    {
      NumberOfCycles=NumberOfEquilibrationCycles;
      if (NumberOfCycles!=0) printf(" Start Equilibration\n");
    }
    else
    {
      if(NumberOfCycles!=0) printf(" Start Production\n");
      NumberOfCycles=NumberOfProductionCycles;
    }
        
    NumberOfAttempts=0;
    NumberOfAcceptedMoves=0;

    // intialize the subroutine that adjust the maximum displacement
    Adjust();

    for(j=0;j<NumberOfCycles;j++)
    {
      // attempt to displace a particle
      for(k=0;k<NumberOfDisplacementsPerCycle;k++)
        Mcmove();

      if(i==PRODUCTION)
      {
        // sample averages
        if((j%SamplingFrequency)==0)
        {
          EnergySquaredSum+=SQR(RunningEnergy);
          EnergySum+=RunningEnergy;
          EnergyCount+=1.0;

          Sample(j,RunningEnergy,RunningVirial,&Pressure,FilePtr);
          PressureSum+=Pressure;
          PressureCount+=1.0;

          // calculate the chemical potential:
          // do 10 trials
          // calculate the average of [exp(-Beta*Energy)]
          // you can use the subroutine EnergyParticle for this

          for(k=0;k<10;k++)
          {
            // start modification


            // end modification
          }
        }
      }

      if((j%20)==0) WritePdb(FilePtrMovie);

      if((j%(NumberOfCycles/5))==0)
      {
        printf("======>> Done %d out of %d\n",j,NumberOfCycles);

        // write intermediate configuration to file
        Store();

        // adjust maximum displacements
        Adjust();
      }
    }
    if(NumberOfCycles!=0)
    {
      if (NumberOfAttempts!=0) 
      {
        printf("Number Of Att. To Displ. A Part.  : %d\n",NumberOfAttempts);
        printf("Success: %d (%lf)\n",NumberOfAcceptedMoves,100.0*NumberOfAcceptedMoves/NumberOfAttempts);
      }

      // test total energy
      EnergySystem();

      if(fabs(TotalEnergy-RunningEnergy)>1.0e-6)
        printf("######### Problems Energy ################\n");
      if(fabs(TotalVirial-RunningVirial)>1.0e-6)
        printf("######### Problems Virial ################\n");

      printf(" Total Energy End Of Simulation    : %lf\n",TotalEnergy);
      printf("       Running Energy              : %lf\n",RunningEnergy);
      printf("       Difference                  : %lf\n",TotalEnergy-RunningEnergy);
      printf(" Total Virial End Of Simulation    : %lf\n",TotalVirial);
      printf("       Running Virial              : %lf\n",RunningVirial);
      printf("       Difference                  : %lf\n",TotalVirial-RunningVirial);

      // print chemical potential and pressure
      if(i==PRODUCTION)
      {
        printf("Average Pressure                  : %lf\n",
          PressureSum/PressureCount);
        printf("Chemical Potential                : %lf\n",
          -log((ChemicalPotentialSum/ChemicalPotentialCount)*(CUBE(Box)/NumberOfParticles))/Beta);
        printf("Heat capacity                     : %lf\n",
          (EnergySquaredSum/EnergyCount)-SQR(EnergySum/EnergyCount));
      }
    }
  }
  Store();

  fclose(FilePtrMovie);
  fclose(FilePtr);

  return 0;
}
